File: translated/f-to-c/DRB004-antidep2-var-yes.c
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.

SPDX-License-Identifier: (BSD-3-Clause)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//Two nested loops with loop-carried anti-dependence on the outer level.
//This is a variable-length array version in F95.
//Data race pair: a[i][j]@55:13:W vs. a[i+1][j]@55:31:R

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char* argv[])
{
    int i, j, len;
    float** a;
    len = 1000;

    if (argc == 1) {
        printf("No command line arguments provided.\n");
    }

    if (argc >= 2) {
        len = atoi(argv[1]);
        if (len <= 0) {
            printf("Error, invalid integer value.\n");
        }
    }

    // Allocate 2D array
    a = (float**)malloc(len * sizeof(float*));
    for (i = 0; i < len; i++) {
        a[i] = (float*)malloc(len * sizeof(float));
    }

    for (i = 0; i < len; i++) {
        for (j = 0; j < len; j++) {
            a[i][j] = 0.5;
        }
    }

    #pragma omp parallel for private(j)
    for (i = 0; i < len-1; i++) {
        for (j = 0; j < len; j++) {
            a[i][j] = a[i][j] + a[i+1][j];
        }
    }

    printf("a(10,10) = %f\n", a[9][9]);

    // Free 2D array
    for (i = 0; i < len; i++) {
        free(a[i]);
    }
    free(a);

    return 0;
} 

File: translated/f-to-go-automodel-bad/DRB004-antidep2-var-yes.go
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
// and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

//Two nested loops with loop-carried anti-dependence on the outer level.
//This is a variable-length array version in F95.
//Data race pair: a[i][j]@55:13:W vs. a[i+1][j]@55:31:R

package main

import (
	"fmt"
	"os"
	"runtime"
	"strconv"
	"sync"
)

func main() {
	var i, j, length int
	var argCount int
	var args []string
	var a [][]float64

	length = 1000

	argCount = len(os.Args) - 1
	if argCount == 0 {
		fmt.Println("No command line arguments provided.")
	}

	args = os.Args[1:]

	if argCount >= 1 {
		var rdErr error
		length, rdErr = strconv.Atoi(args[0])
		if rdErr != nil {
			fmt.Println("Error, invalid integer value.")
		}
	}

	a = make([][]float64, length)
	for i := range a {
		a[i] = make([]float64, length)
	}

	for i = 1; i <= length; i++ {
		for j = 1; j <= length; j++ {
			a[i-1][j-1] = 0.5
		}
	}

	//$omp parallel do private(j)
	var wg sync.WaitGroup
	numCPU := runtime.NumCPU()
	chunkSize := (length - 1) / numCPU
	if chunkSize < 1 {
		chunkSize = 1
	}

	for start := 1; start <= length-1; start += chunkSize {
		end := start + chunkSize - 1
		if end > length-1 {
			end = length - 1
		}
		wg.Add(1)
		go func(start, end int) {
			defer wg.Done()
			for i := start; i <= end; i++ {
				for j := 1; j <= length; j++ {
					a[i-1][j-1] = a[i-1][j-1] + a[i][j-1]
				}
			}
		}(start, end)
	}
	wg.Wait()
	//$omp end parallel do

	fmt.Printf("a(10,10) = %f\n", a[9][9])
} 